File: Xahla-API-Core/src/xahla/client/graphics/Window.java
package xahla.client.graphics;

import static org.lwjgl.glfw.Callbacks.glfwFreeCallbacks;
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL11.glClearColor;
import static org.lwjgl.system.MemoryStack.stackPush;
import static org.lwjgl.system.MemoryUtil.NULL;

import java.nio.IntBuffer;

import org.joml.Vector2i;
import org.lwjgl.BufferUtils;
import org.lwjgl.glfw.GLFWErrorCallback;
import org.lwjgl.glfw.GLFWVidMode;
import org.lwjgl.opengl.GL;
import org.lwjgl.opengl.GL11;
import org.lwjgl.system.MemoryStack;

import xahla.client.input.Input;
import xahla.context.ClientContext;
import xahla.core.IAppCore;

/**
 * This class handles GLFW.
 * It display a window and manage user input.
 * 
 * @author Cochetooo
 * @version 1.1
 */
public class Window implements IAppCore {
	
	private long window;
	private ClientContext Context;
	
	private GLFWConfiguration config;
	
	private Vector2i windowSize;
	
	/**
	 * @param Context 	The program context.
	 */
	public Window(ClientContext Context) {
		this.Context = Context;
		
		this.windowSize = new Vector2i(
			Context.getConfigInt("Rendering", "initial_width"),
			Context.getConfigInt("Rendering", "initial_height")
			);
		
		boolean resizable = Context.getConfigBool("Rendering", "resizable");
		
		this.config = new GLFWConfiguration(resizable ? GLFW_TRUE : GLFW_FALSE, windowSize.x, windowSize.y, "Kina Siern 080920");
	}

	/** Create the window and set callback for input. */
	@Override
	public void init() {
		GLFWErrorCallback.createPrint(System.err).set();
		
		if (!glfwInit())
			throw new IllegalStateException("Unable to initialize GLFW");
		
		glfwDefaultWindowHints();
		glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
		
		glfwWindowHint(GLFW_RESIZABLE, config.isResizable());
		
		Vector2i size = config.getSize();
		window = glfwCreateWindow(size.x, size.y, config.getTitle(), NULL, NULL);
		if (window == NULL)
			throw new RuntimeException("Failed to create the GLFW window");
		
		try (MemoryStack stack = stackPush()) {
			IntBuffer pWidth = stack.mallocInt(1);
			IntBuffer pHeight = stack.mallocInt(1);
			
			glfwGetWindowSize(window, pWidth, pHeight);
			GLFWVidMode vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor());
			
			glfwSetWindowPos(
				window,
				(vidmode.width() - pWidth.get(0)) / 2,
				(vidmode.height() - pHeight.get(0)) / 2
			);
		}
		
		glfwMakeContextCurrent(window);
		glfwShowWindow(window);
		
		Input.init(window);
		
		glfwSetKeyCallback(window, Input.keyboard);
		glfwSetMouseButtonCallback(window, Input.mouse);
		
		GL.createCapabilities();
		glClearColor(0.f, .0f, .0f, .0f);
		
		GL11.glViewport(0, 0, windowSize.x, windowSize.y);
	}
	
	/** Update the window and the input. */
	@Override
	public void update() {
		if (glfwWindowShouldClose(window))
			Context.request("dispose");
		
		Input.update();
	}
	
	/** Swap display buffers. */
	@Override
	public void render() {
		glfwSwapBuffers(window);
	}
	
	/** Free callbacks and destroy the window. */
	@Override
	public void dispose() {
		glfwFreeCallbacks(window);
		glfwDestroyWindow(window);
		
		glfwTerminate();
		glfwSetErrorCallback(null).free();
	}
	
	/** @param b Set display vertical synchronisation. */
	public void setVSync(boolean b) {
		glfwSwapInterval(b ? 1 : 0);
	}
	
	/** @param title Set the window title. */
	public void changeTitle(String title) {
		glfwSetWindowTitle(window, title);
	}
	
	/** @return The Window dimension as a Vector. */
	public final Vector2i getWindowDimension() {
		IntBuffer w, h;
		
		try (MemoryStack stack = stackPush()) {
			w = BufferUtils.createIntBuffer(1);
			h = BufferUtils.createIntBuffer(1);
			glfwGetWindowSize(window, w, h);
			
			return windowSize.set(w.get(0), h.get(0));
		}
	}
	
	/** @return The Window properties such as width, height, etc. */
	public final GLFWConfiguration getConfiguration() { return config; }
}

File: Render/src/main/kotlin/graphics/Window.kt
package graphics

import XHR_OPENGL
import XHR_VULKAN
import app
import config
import ClientContext
import input.input
import org.joml.Vector2i
import org.lwjgl.glfw.Callbacks.glfwFreeCallbacks
import org.lwjgl.glfw.GLFW.*
import org.lwjgl.glfw.GLFWErrorCallback
import org.lwjgl.glfw.GLFWWindowSizeCallback
import org.lwjgl.opengl.GL
import org.lwjgl.opengl.GL11.glClearColor
import org.lwjgl.opengl.GL11.glViewport
import org.lwjgl.system.MemoryStack.stackPush
import org.lwjgl.system.MemoryUtil.NULL
import templates.IEngine
import utils.LogLevel
import utils.XH_STATUS_GLFW_ERROR
import utils.logger
import java.lang.IllegalStateException
import java.lang.RuntimeException
import kotlin.math.pow

data class GLFWConfiguration(val resizable: Int, val fullscreen: Int, val width: Int, val height: Int,
                             val title: String, val colorBufferBits: Int, val floating: Int, val decoration: Int, val msaa: Int,
                             val centerCursor: Int, val vSync: Int) {
    fun getSize() = Vector2i(width, height)
}

class XHR_Window(val context: ClientContext) : IEngine {
    var window: Long = 0L
        private set

    val dimension = Vector2i(config("window.screenWidth") as Int, config("window.screenHeight") as Int)

    val config: GLFWConfiguration = GLFWConfiguration(
        if (config("window.resizable") as Boolean) GLFW_TRUE else GLFW_FALSE,
        if (config("window.fullscreen") as Boolean) GLFW_TRUE else GLFW_FALSE,
        dimension.x, dimension.y,
        config("window.title") as String,
        config("gl.colorBufferBits") as Int,
        if (config("window.floating") as Boolean) GLFW_TRUE else GLFW_FALSE,
        if (config("window.decorated") as Boolean) GLFW_TRUE else GLFW_FALSE,
        config("gl.multisampling") as Int,
        if (config("window.centerCursor") as Boolean) GLFW_TRUE else GLFW_FALSE,
        if (config("gl.vsync") as Boolean) GLFW_TRUE else GLFW_FALSE
    )

    override fun onInit() {
        GLFWErrorCallback.createPrint(logger().printer).set()

        if (!glfwInit())
            logger().throwException("Unable to initialize GLFW", IllegalStateException(),
                classSource = "XHR_Window", statusCode = XH_STATUS_GLFW_ERROR)

        glfwDefaultWindowHints()
        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE)
        glfwWindowHint(GLFW_RESIZABLE, config.resizable)
        glfwWindowHint(GLFW_DECORATED, config.decoration)
        glfwWindowHint(GLFW_CENTER_CURSOR, config.centerCursor)
        glfwWindowHint(GLFW_FLOATING, config.floating)

        glfwWindowHint(GLFW_RED_BITS, config.colorBufferBits / 4)
        glfwWindowHint(GLFW_GREEN_BITS, config.colorBufferBits / 4)
        glfwWindowHint(GLFW_BLUE_BITS, config.colorBufferBits / 4)
        glfwWindowHint(GLFW_ALPHA_BITS, config.colorBufferBits / 4)

        if (config.msaa > 0)
            glfwWindowHint(GLFW_SAMPLES, 2.0.pow(config.msaa).toInt())

        window = glfwCreateWindow(dimension.x, dimension.y, config.title, NULL, NULL)

        if (window == NULL)
            logger().throwException("Failed to create the GLFW window", RuntimeException(),
                classSource = "XHR_Window", statusCode = XH_STATUS_GLFW_ERROR)

        stackPush().use {
            val pWidth = it.mallocInt(1)
            val pHeight = it.mallocInt(1)

            glfwGetWindowSize(window, pWidth, pHeight)
            val vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor())
                ?: logger().throwException("Failed to retrieve a screen monitor", RuntimeException(),
                    classSource = "XHR_Window", statusCode = XH_STATUS_GLFW_ERROR)

            glfwSetWindowPos(
                window,
                (vidmode.width() - pWidth[0]) / 2,
                (vidmode.height() - pHeight[0]) / 2
            )
        }

        glfwMakeContextCurrent(window)
        glfwShowWindow(window)

        input().init(window)

        glfwSetKeyCallback(window, input().keyboard)
        glfwSetMouseButtonCallback(window, input().mouse)
        glfwSetWindowSizeCallback(window, object: GLFWWindowSizeCallback() {
            override fun invoke(win: Long, width: Int, height: Int) {
                if (win != window) return

                onResize()
            }
        })

        when (config("window.engine")) {
            XHR_OPENGL -> {
                GL.createCapabilities()
                glClearColor(.0f, .0f, .0f, .0f)
                glViewport(0, 0, dimension.x, dimension.y)
            }
            XHR_VULKAN -> {

            }
        }

        setVSync(config.vSync)

        logger().internal_log("Window created successfully!", LogLevel.FINE, "Window")
    }

    override fun onUpdate() {
        if (glfwWindowShouldClose(window))
            app().stop()

        input().onUpdate()
    }

    override fun onPostRender() {
        glfwSwapBuffers(window)
    }

    override fun onResize() {
        stackPush().use {
            val w = it.mallocInt(1)
            val h = it.mallocInt(1)
            glfwGetWindowSize(window, w, h)

            dimension.set(w[0], h[0])
        }

        glViewport(0, 0, dimension.x, dimension.y)
        context.onResize()
    }

    override fun onDispose() {
        glfwFreeCallbacks(window)
        glfwDestroyWindow(window)

        glfwTerminate()
        glfwSetErrorCallback(null)?.free()
    }

    fun setCursorMode(mode: Int) = glfwSetInputMode(window, GLFW_CURSOR, mode)
    fun setVSync(interval: Int) = glfwSwapInterval(interval)
    fun setWindowTitle(title: String) = glfwSetWindowTitle(window, title)

}